### PR TITLE
New extended tutorial for test jobs (infra)

### PR DESCRIPTION
**Pull request by Hook25**
## Description

We are currently lacking a tutorial to get started with test authoring. 

This is an attempt to give a comprehensive introduction to everything one may want to know about automated tests via a tutorial that guides the user through creating a new provider with:
- automated tests
- templates
- resources
- inter-test dependencies
- binaries to build via mange build

This tutorial is only about tests, so the absence of test plans is intentional. This is meant to be the first in a chain of "advanced/extended" tutorials (this is why it also contains the instructions to create the provider and provision Checkbox from source)

## Resolved issues

Fixes: https://warthogs.atlassian.net/browse/CHECKBOX-761

## Documentation

N/A

## Tests

N/A

### CODE REVIEW

File: docs/tutorial/writing-tests/test-case.rst [hunk 1]
```diff
@@ -0,0 +1,1105 @@
...
+
+Before running the test we have to define a Checkbox configuration. Note that
+if we were using a test plan, we could run it with a launcher, but the
+``run`` command doesn't take a launcher parameter, so we have to use a
+configuration file. Place the following in ``~/.config/checkbox.conf``.
+
+.. code-block:: ini
+
+    [environment]
+    ACCEPTABLE_BYTES_PER_SECOND_SPEED=60000000
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:233]: I see `60000000` here, `600` as the default in the job, and I'm a bit confused, especially as `-Y` is defined as

```
       -Y, --speed-limit <speed>
              If a transfer is slower than this given speed (in bytes per sec‐
              ond) for speed-time seconds it gets aborted. speed-time  is  set
              with -y, --speed-time and is 30 if not set.
```

File: docs/tutorial/writing-tests/test-case.rst [hunk 2]
```diff
@@ -0,0 +1,1105 @@
...
+If you expect a variable to be set and it is not, always fail the test stating
+what variable you needed and what it was for. If you decide to use a default
+value, always output the value the test is going to use in the test log so that
+when you have to investigate why something went wrong, it is trivial to
+reproduce the tests with the parameters that may have made it fail.
+
+Resources
+=========
+
+Before even thinking to test if we are connected to the internet a wise
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:273]:
```suggestion
Before even thinking to test if we are connected to the Internet a wise
```

File: docs/tutorial/writing-tests/test-case.rst [hunk 3]
```diff
@@ -0,0 +1,1105 @@
...
+question to ask would be: do we even have a network interface? Lets create a
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:274]:
```suggestion
question to ask would be: do we even have a network interface? Let's create a
```

File: docs/tutorial/writing-tests/test-case.rst [hunk 4]
```diff
@@ -0,0 +1,1105 @@
...
+resource job to fetch this information.
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:275]: What is a "resource job"? Either link to doc, or explain in one sentence (good luck).

File: docs/tutorial/writing-tests/test-case.rst [hunk 5]
```diff
@@ -0,0 +1,1105 @@
...
+
+
+    (checkbox_venv) > python3 manage.py validate
+    [...]
+    error: ../base/units/submission/packaging.pxu:3: field 'Depends', clashes with 1 other unit, look at: ../base/units/submission/packaging.pxu:1-3, units/extended_tutorial.pxu:1-4
+    Validation of provider tutorial has failed
+
+Opening the file that the validator complains about, you will notice that the
+jq dependency is already required by a base provider test. We can rely on the
+base provider, so we can safely remove this dependency from our provider.
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:314]: This gets difficult pretty quickly, mixing packaging meta data units with resource jobs. Maybe this can be skipped for now, given that the base provider includes it (and, actually, it's not required to run in a venv)?

File: docs/tutorial/writing-tests/test-case.rst [hunk 6]
```diff
@@ -0,0 +1,1105 @@
...
+
+.. warning::
+   If you don't have ``jq`` installed on your machine, install it now else you
+   won't be able to follow the next steps. You can install it either via
+   ``sudo snap install jq`` or ``sudo apt install jq``
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:319]:
```suggestion
   ``sudo snap install jq`` or ``sudo apt install jq``.
```

File: docs/tutorial/writing-tests/test-case.rst [hunk 7]
```diff
@@ -0,0 +1,1105 @@
...
+
+Now that we have this new resource lets run it to see what the output is
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:321]:
```suggestion
Now that we have this new resource let's run it to see what the output is
```

File: docs/tutorial/writing-tests/test-case.rst [hunk 8]
```diff
@@ -0,0 +1,1105 @@
...
+    interface: veth993f2cd0
+    link_info_kind: veth
+    link_type: ether
+
+    interface: tun0
+    link_info_kind: tun
+    link_type: none
+
+We now add a ``requires:`` constraint to our jobs so that, if no interface
+that could possibly connected to the internet is on the machine, we can
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:361]:
```suggestion
that could possibly connect to the Internet is on the machine, we can
```

File: docs/tutorial/writing-tests/test-case.rst [hunk 9]
```diff
@@ -0,0 +1,1105 @@
...
+skip them instead of failing.
+
+.. code-block:: none
+    :emphasize-lines: 4,5
+
+    id: network_available
+    flags: simple
+    _summary: Test that the internet is reachable
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:369]:
```suggestion
    _summary: Test that the Internet is reachable
```

File: docs/tutorial/writing-tests/test-case.rst [hunk 10]
```diff
@@ -0,0 +1,1105 @@
...
+    requires:
+      network_iface_info.link_type == "ether"
+    command:
+      ping -c 1 1.1.1.1
+
+If we now run the ``network_available`` test, Checkbox will automatically pull
+also ``network_iface_info``. Note that this only happens because both are in
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:376]:
```suggestion
If we now run the ``network_available`` test, Checkbox will also automatically
pull ``network_iface_info``. Note that this only happens because both are in
```

File: docs/tutorial/writing-tests/test-case.rst [hunk 11]
```diff
@@ -0,0 +1,1105 @@
...
+
+.. code-block:: none
+
+    (checkbox_venv) > checkbox-cli run com.canonical.certification::network_available
+    ===========================[ Running Selected Jobs ]============================
+    =========[ Running job 1 / 2. Estimated time left (at least): 0:00:00 ]=========
+    ----------------[ Fetches information of all network intefaces ]----------------
+    [...]
+    =========[ Running job 2 / 2. Estimated time left (at least): 0:00:00 ]=========
+    --------------------[ Test that the internet is reachable ]---------------------
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:387]:
```suggestion
    --------------------[ Test that the Internet is reachable ]---------------------
```

File: docs/tutorial/writing-tests/test-case.rst [hunk 12]
```diff
@@ -0,0 +1,1105 @@
...
+    [...]
+    ==================================[ Results ]===================================
+     ☑ : Fetches information of all network intefaces
+     ☑ : Test that the internet is reachable
+
+Are we done then? Almost, there are a few issues with our resource job. The
+first and most relevant is that the ``resource`` constraint we have written
+seems to work, but if we analyze the output what we have written actually
+over-matches (as ``veth993f2cd0`` is also an ``ether`` device, but it is not a
+valid interface to use to connect to the internet). We can easily fix this by
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:397]:
```suggestion
valid interface to use to connect to the Internet). We can easily fix this by
```

File: docs/tutorial/writing-tests/test-case.rst [hunk 13]
```diff
@@ -0,0 +1,1105 @@
...
+updating the expression as follows but take note of what happened.
+
+.. warning::
+    It is actually difficult to write a significant resource expressions. This
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:401]:
```suggestion
    It is actually difficult to write a significant resource expression. This
```

File: docs/tutorial/writing-tests/test-case.rst [hunk 14]
```diff
@@ -0,0 +1,1105 @@
...
+    time we got "lucky", and we could notice the mistake on our own machine, but
+    this may not be the always the case. In general make your resource
+    expressions as restrictive as possible.
+
+.. code-block:: none
+
+    id: network_available
+    [...]
+    requires:
+      (network_iface_info.link_info_kind == "" and network_iface_info.link_type == "ether")
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:411]: You prefer to show them the best way to write resource expressions, instead of the following (albeit easier to read for a human)?

```suggestion
      network_iface_info.link_info_kind == ""
      network_iface_info.link_type == "ether"
```

File: docs/tutorial/writing-tests/test-case.rst [hunk 15]
```diff
@@ -0,0 +1,1105 @@
...
+
+The second issue is harder to fix. Checkbox is currently built for a multitude
+of Ubuntu versions, including 16.04. If we inspect the 16.04
+`manual <https://manpages.ubuntu.com/manpages/xenial/man8/ip.8.html>`_ of the
+``ip`` command we notice one thing: the version shipped with Xenial doesn't support
+the ``--json`` flag.
+
+.. warning::
+    When you use a pre-installed package, always check if all versions support
+    your use case and if there is a version available for all target versions
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:421]:
```suggestion
    your use case and if there is a version available for all target versions.
```

File: docs/tutorial/writing-tests/test-case.rst [hunk 16]
```diff
@@ -0,0 +1,1105 @@
...
+    interface: wlan0
+    [...]
+
+    (checkbox_venv) > checkbox-cli run com.canonical.certification::device | grep -C 15 enp
+    [...]
+    category: NETWORK
+    interface: enp5s0
+    [...]
+
+Lets propagate this newfound knowledge over to our ``requires`` constraint:
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:453]:
```suggestion
Let's propagate this newfound knowledge over to our ``requires`` constraint:
```

**Review by pieqq — CHANGES_REQUESTED** [review]
Very cool stuff indeed!

I only reviewed until the `Template Jobs` section.

Writing tutorials is very hard. It's tempting to try to explain many things at once, but this might become too hard to read for a newcomer. If you feel that you have to explain a concept while explaining another one, this might be better to split this into another section, later.

File: docs/tutorial/writing-tests/test-case.rst [hunk 17]
```diff
@@ -0,0 +1,1105 @@
...
+    environ:
+      ACCEPTABLE_BYTES_PER_SECOND_SPEED
+    command:
+      echo Testing for the limit speed: ${ACCEPTABLE_BYTES_PER_SECOND_SPEED:-600}
+      curl -Y ${ACCEPTABLE_BYTES_PER_SECOND_SPEED:-600} -o /dev/null \
+        https://cdimage.ubuntu.com/ubuntu-mini-iso/noble/daily-live/current/noble-mini-iso-amd64.iso
+
+Before running the test we have to define a Checkbox configuration. Note that
+if we were using a test plan, we could run it with a launcher, but the
+``run`` command doesn't take a launcher parameter, so we have to use a
```
[the remark this answers is not in the document]
Comment (Hook25) [reply docs/tutorial/writing-tests/test-case.rst:227]: Agree, run is so underpower it makes me sad sometimes

File: docs/tutorial/writing-tests/test-case.rst (hunk 1 above)
Comment (Hook25) [reply docs/tutorial/writing-tests/test-case.rst:233]: (ops, -y is 30 by default, I thought it was 1)

File: docs/tutorial/writing-tests/test-case.rst (hunk 4 above)
Comment (Hook25) [reply docs/tutorial/writing-tests/test-case.rst:275]: Link to the doc and explanation is in order, I think this tutorial should assume you went through the basic tutorial and you are a yellow belt checkbox user, ready to start writing

File: docs/tutorial/writing-tests/test-case.rst (hunk 14 above)
Comment (Hook25) [reply docs/tutorial/writing-tests/test-case.rst:411]: This is completely wrong though, this means there is at least a network_iface_info that has no link speed and one that is of type ether. the one in the tutorial means there is a network_iface_info that is of kind empty AND link type ether

File: docs/tutorial/writing-tests/test-case.rst (hunk 5 above)
Comment (Hook25) [reply docs/tutorial/writing-tests/test-case.rst]: I tried to give a progression that goes through the path of a new job developer in a rational progression. I think that here one would ask themselves: well I've used this tool, but how does checkbox know?

File: docs/tutorial/writing-tests/test-case.rst [hunk 18]
```diff
@@ -0,0 +1,1105 @@
...
+===========================
+
+Ok, it worked, but this is not very useful. Lets go back and edit the job to
+actually run a ping command. Replace the command section of the job with
+``ping -c 1 1.1.1.1``, lets also update the summary as follows:
+
+.. code-block:: none
+
+    id: network_available
+    flags: simple
```
[the remark this answers is not in the document]
Comment (Hook25) [reply docs/tutorial/writing-tests/test-case.rst:75]: Added an explanation and a lik to the glossary in a note below the first usage of a flag

File: docs/tutorial/writing-tests/test-case.rst [hunk 19]
```diff
@@ -0,0 +1,1148 @@
...
+                rtt min/avg/max/mdev = 34.980/60.486/142.567/31.077 ms
+                """
+            ).strip()
+            with self.assertRaises(SystemExit):
+                network_available.network_available("wlan0", "0")
+
+.. note::
+   We use ``self.assertTrue(check_output_mock.called)`` instead of
+   ``check_output_mock.assert_called_once()``. The reason is that we have to be
+   compatible (in tests as well!) with python3.5 and
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:948]:
```suggestion
   compatible (in tests as well!) with Python 3.5 and
```

File: docs/tutorial/writing-tests/test-case.rst [hunk 20]
```diff
@@ -0,0 +1,1148 @@
...
+   ``Mock.assert_called_once`` was introduced in Python3.6. If you don't know
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:949]:
```suggestion
   ``Mock.assert_called_once`` was introduced in Python 3.6. If you don't know
```

File: docs/tutorial/writing-tests/test-case.rst [hunk 21]
```diff
@@ -0,0 +1,1148 @@
...
+   when a function was introduced, refer to `the python documentation
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:950]:
```suggestion
   when a function was introduced, refer to `the Python documentation
```

File: docs/tutorial/writing-tests/test-case.rst [hunk 22]
```diff
@@ -0,0 +1,1148 @@
...
+   <https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called_once>`_.
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:951]:
```suggestion
   <https://docs.python.org/3/>`_.
```

File: docs/tutorial/writing-tests/test-case.rst [hunk 23]
```diff
@@ -0,0 +1,1148 @@
...
+   See below the example, if a function was introduced with a specific
+   version, you will find it there.
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:953]:
```suggestion
   For example, if you check the documentation for `Mock.assert_called_once<https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called_once>`_ you will see *Added in version 3.6.*
```

File: docs/tutorial/writing-tests/test-case.rst [hunk 24]
```diff
@@ -0,0 +1,1148 @@
...
+    ----------------------------------------------
+    TOTAL                         25     10    60%
+    (checkbox_venv) > python3 -m coverage report --include=bin/* -m
+    Name                       Stmts   Miss  Cover   Missing
+    --------------------------------------------------------
+    bin/network_available.py      25     10    60%   8-18, 29, 49-52, 56
+    --------------------------------------------------------
+    TOTAL                         25     10    60%
+
+    # You can also get an html report with the following
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:1002]:
```suggestion
    # You can also get an HTML report with the following
```

File: docs/tutorial/writing-tests/test-case.rst [hunk 25]
```diff
@@ -0,0 +1,1148 @@
...
+    # it is very convenient as you can see file per file what lines are covered
+    # in
+    (checkbox_venv) > python3 -m coverage html
+
+As you can see we are way below the coverage target but this is difficult to
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:1007]:
```suggestion
As you can see we are way below the coverage target (90%) but this is difficult to
```

File: docs/tutorial/writing-tests/test-case.rst [hunk 26]
```diff
@@ -0,0 +1,1148 @@
...
+                64 bytes from 1.1.1.1: icmp_seq=2 ttl=53 time=143 ms
+
+                --- 1.1.1.1 ping statistics ---
+                2 packets transmitted, 2 received, 0% packet loss, time 170ms
+                rtt min/avg/max/mdev = 34.980/60.486/142.567/31.077 ms
+                """
+            ).strip()
+            network_available.main(["--threshold", "20", "wlan0"])
+            self.assertTrue(check_output_mock.called)
+
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:1031]: You can add a conclusion sentence for this section like

> We are now reaching *X%* of test coverage, this is good enough for what we want to achieve.
> 
> Having unit tests for Python scripts used in the Checkbox jobs will help you and others maintain the scripts and prevent regressions in the future.

(or something along the way)

File: docs/tutorial/writing-tests/test-case.rst [hunk 27]
```diff
@@ -0,0 +1,1148 @@
...
+Now we can go back to the root of the provider and use ``manage.py`` to compile
+our test file:
+
+.. code:: none
+
+    (checkbox_venv) > ./manage.py build
+    cc -Wall -pedantic ../../src/vfork_memory_share_test.c -o vfork_memory_share_test
+    # The following step is not necessary when you install a provider
+    # but
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:1106]: is there something missing here, or can the comment be removed?

File: docs/tutorial/writing-tests/test-case.rst [hunk 28]
```diff
@@ -0,0 +1,1148 @@
...
+    ... 8< -------------------------------------------------------------------------
+    Child set the variable, vfork shares the memory
+    ------------------------------------------------------------------------- >8 ---
+    Outcome: job passed
+    Finalizing session that hasn't been submitted anywhere: checkbox-run-2024-08-08T13.35.24
+    ==================================[ Results ]===================================
+     ☑ : Check that vfork syscall shares the memory between parent and child
+
+.. warning::
+   Checkbox is delivered for many platforms so be mindful of what you include
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:1138]:
```suggestion
   Checkbox is delivered for many platforms (x86, ARM, etc.) so be mindful of what you include
```

File: docs/tutorial/writing-tests/test-case.rst [hunk 29]
```diff
@@ -0,0 +1,1148 @@
...
+   in the ``src/`` directory, especially if you plan to contribute the test
+   upstream. It must be compatible with all architectures we build for, debian
```
Comment (pieqq) [inline docs/tutorial/writing-tests/test-case.rst:1140]:
```suggestion
   upstream. It must be compatible with all architectures we build for, Debian
```

**Review by pieqq — CHANGES_REQUESTED** [review]
Brilliant work, really! This is a long tutorial, and you introduce a lot of concepts that are unique to Checkbox in a easy-to-follow way. Congratulations!

You even include obscure stuff like the ability to compile source code... impressive!

I've made a bunch of suggestions and typo fixes as well as other general comments. Once these land, we can merge this PR.

Thanks again, great work!

File: docs/tutorial/writing-tests/test-case.rst [hunk 30]
```diff
@@ -0,0 +1,1148 @@
...
+  to pass for an external reason (for example, the ethernet port may work but
+  it is not currently plugged in)
+
+Dealing with complexity - Python
+================================
+
+The ``network_available`` test that we have created during this tutorial is
+very simple but, in the real world things are not as simple. For example,
+right now we are only pinging once from the test, if the ping goes through
+we call that a success, else a failure. This works in our simple scenario while
```
Comment (tang-mm) [inline docs/tutorial/writing-tests/test-case.rst:732]:
```suggestion
the test is considered successful; otherwise, it's a failure. This works in our simple scenario while
```
a bit like a mixture of Python and English :)

File: docs/tutorial/writing-tests/test-case.rst [hunk 31]
```diff
@@ -0,0 +1,1148 @@
...
+        ping_test(args.interface)
+
+
+    if __name__ == "__main__":
+        main()
+
+.. note::
+    A few important things to notice about the script:
+
+    #. We use Black to format all tests and source files in Checkbox with a custom config: ``line-length = 79``.
```
Comment (tang-mm) [inline docs/tutorial/writing-tests/test-case.rst:796]:
```suggestion
    - We use Black to format all tests and source files in Checkbox with a custom config: ``line-length = 79``.
```
Ordering is not important in the list, so bullets would be better (which is used in a note below)

File: docs/tutorial/writing-tests/test-case.rst [hunk 32]
```diff
@@ -0,0 +1,1148 @@
...
+In Checkbox we have a coverage requirement for new pull requests.
+This is to ensure that new contributions do not add source paths that are not
+explored in testing and therefore easy to break down the line with any change.
+
+If you want to collect the coverage of your contribution you can run the
+following:
+
+.. code-block:: none
+
+    (checkbox_venv) > python3 -m coverage run manage.py test -u
```
Comment (tang-mm) [inline docs/tutorial/writing-tests/test-case.rst:988]: Looks like `coverage` has to be installed separately. so perhaps  `pip install coverage` is needed before this step

File: docs/tutorial/writing-tests/test-case.rst (hunk 27 above)
Comment (tang-mm) [inline docs/tutorial/writing-tests/test-case.rst:1106]: is it intentional?

File: docs/tutorial/index.rst [hunk 33]
```diff
@@ -42,3 +42,8 @@ Extended tutorial
 
 Once you have completed the core tutorial, the extended optional tutorial
 sections can be followed in any order - they don’t depend on each other.
+
+.. toctree::
+   :maxdepth: 1
+
+  writing-tests/test-case
```
Comment (tang-mm) [inline docs/tutorial/index.rst:49]: Suggestion:
I wonder if it'd be helpful to split the sections into new pages and list them in sequence, so that the document is easier to navigate and maintain. (like in `docs/reference/units/`)
Also, some of the sections are useful in other scenarios as well (e.g. defining resources, unit testing, complexity.), so it'd be great to have them appear in the table of contents panel for discoverability.

Comment (tang-mm) [reply docs/tutorial/index.rst:49]: There's also an option to show the ordered number before the title without changing the files.
```
.. toctree::
   :maxdepth: 1
   :numbered:
```

for example:
image

File: docs/tutorial/writing-tests/test-case.rst [hunk 34]
```diff
@@ -0,0 +1,1148 @@
...
+    (checkbox_venv) > cd ../providers/resource
+    (checkbox_venv) > python3 manage.py develop
+
+.. note::
+    Remember to activate the virtual environment! You can also create an alias
+    in your ``~/.bashrc`` to enable it when you need it.
+
+Creating a new provider
+=======================
+
```
[the remark this answers is not in the document]
Comment (tang-mm) [reply docs/tutorial/writing-tests/test-case.rst:42]: For cross-reference: In the explanation of Provider `explanation/understanding.rst`, we said "See tutorials on how to create a provider from scratch". It needs to be updated to link to this new tutorial :)

**Review by tang-mm — CHANGES_REQUESTED** [review]
Kudos for devising such a thorough tutorial!!
It's easy to follow, clearly explains what/why we are doing so in each step, and all the tips/suggestions are very useful. I left a bunch of comments but most of them are nitpicks. Feel free to adapt as needed


And a thought for future consideration (for me):
This tutorial is a good candidate for documentation testing - all the code samples are executable and already in the repository. Instead of duplicating the code in the doc, maybe we can include the snippet from source files into the document, so the tutorial can be validated as well.

File: docs/tutorial/writing-tests/test-case.rst [hunk 35]
```diff
@@ -0,0 +1,1148 @@
...
+Let's create a new Checkbox provider by using the Checkbox sub-command
+``startprovider``.
```
[the remark this answers is not in the document]
Comment (Hook25) [reply docs/tutorial/writing-tests/test-case.rst:44]: Disregarding this as it can (should) be done in any directory. Most likely not in the `checkbox/providers` one, we don't really accept new providers in the repo that way.

File: docs/tutorial/writing-tests/test-case.rst [hunk 36]
```diff
@@ -0,0 +1,1148 @@
...
+    ------------------------------------------------------------------------- >8 ---
+    Outcome: job passed
+    Finalizing session that hasn't been submitted anywhere: checkbox-run-2024-08-06T14.17.23
+    ==================================[ Results ]===================================
+     ☑ : Test that the network speed is acceptable
+
+
+.. warning::
+
+    Don't assume that a Checkbox job will inherit any environment variable from
```
[the remark this answers is not in the document]
Comment (Hook25) [reply docs/tutorial/writing-tests/test-case.rst:296]: (partially false, but it is better to scare than to explain the edge cases in this situation!)

File: docs/tutorial/writing-tests/test-case.rst [hunk 37]
```diff
@@ -0,0 +1,1148 @@
...
+    flags: simple
+    _summary: Test that the Internet is reachable
+    requires:
+      network_iface_info.link_type == "ether"
+    command:
+      ping -c 1 1.1.1.1
+
+If we now run the ``network_available`` test, Checkbox will also automatically
+pull ``network_iface_info``. Note that this only happens because both are in
+the same namespace.
```
[the remark this answers is not in the document]
Comment (Hook25) [reply docs/tutorial/writing-tests/test-case.rst:420]: probably not

File: docs/tutorial/writing-tests/test-case.rst [hunk 38]
```diff
@@ -0,0 +1,1148 @@
...
+  executed via run
+- Templates can't be executed via run using their ``template-id``
+
+We can easily solve the situation in this example by manually pulling the
+dependency and using the explicit id of the job that will be generated or a
+regex:
+
+.. code-block:: none
+
+    (checkbox_venv) > checkbox-cli run com.canonical.certification::network_iface_info "com.canonical.certification::network_available_wlan0"
```
[the remark this answers is not in the document]
Comment (Hook25) [reply docs/tutorial/writing-tests/test-case.rst:568]: it is not, but I've noticed the namespace is wrong, I will fix that

File: docs/tutorial/writing-tests/test-case.rst [hunk 39]
```diff
@@ -0,0 +1,1148 @@
...
+     ☑ : Test that the internet is reachable via enp2s0f0
+     ☑ : Test that the internet is reachable via enp5s0
+     ☑ : Test that the internet is reachable via wlan0
+     ☑ : Test that the internet is reachable via lxdbr0
+     ☑ : Test that the internet is reachable via vetha6dd5923
+
+This is a quick and dirty solution that can be handy if you want to run a test
+and you can manually resolve the dependency chain that is not resolved by
+Checkbox but this can be, in practice, often hard or impossible.
+For a more principled solution see the the Test Plan Tutorial section.
```
[the remark this answers is not in the document]
Comment (Hook25) [reply docs/tutorial/writing-tests/test-case.rst:589]: we have to create it! it is not created yet

File: docs/tutorial/writing-tests/test-case.rst (hunk 22 above)
Comment (Hook25) [reply docs/tutorial/writing-tests/test-case.rst:951]: I was linking the topic I was talking about!

File: docs/tutorial/writing-tests/test-case.rst (hunk 26 above)
Comment (Hook25) [reply docs/tutorial/writing-tests/test-case.rst:1035]: I think the paragraph introduction above is pretty clear

File: docs/tutorial/index.rst (hunk 33 above)
Comment (Hook25) [reply docs/tutorial/index.rst:49]: I think this tutorial is better takes as a full, 1 session lesson on checkbox tests, so I would rather keep it as a single big document
